server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -78,6 +78,35 @@ public class TransportGetSnapshotsAction extends TransportMasterNodeAction<GetSn
 
     private final RepositoriesService repositoriesService;
 
+    /*
+     * [NOTE ON THREADING]
+     *
+     * This action runs three kinds of task:
+     *
+     * 1. looping through all the SnapshotInfo instances that we can construct from in-memory data (in-progress snapshots and the
+     * verbose=false case), adding them to the results.
+     * 2. adding each remaining SnapshotInfo instance to the results after loading it from the repository.
+     * 3. processing the results into the final response.
+     *
+     * It uses two pools:
+     *
+     * SNAPSHOT_META:: Intended for loading metadata from the repository. Fairly large, so not appropriate for CPU-intensive tasks. Also
+     * used by other snapshot operations so we must not use it for long-running work or spam its queue. This is used only for the type-2
+     * tasks above, since it's the pool that loads the SnapshotInfo from the repository in the first place, and we avoid spamming its
+     * queue for those tasks with GetSnapshotInfoExecutor.
+     *
+     * MANAGEMENT:: Intended for other management-related work. Small, and not used for critical production activites, so it's ok to do
+     * long-running or CPU-intensive things here. This is used for type-1 and type-3 tasks which both loop over many SnapshotInfo instances
+     * at once.
+     *
+     * The mechanism for achieving this threading model is a little subtle. We start processing each repository on MANAGEMENT either because
+     * that's where masterOperation() runs or because we fork back to MANAGEMENT after getRepositoryData(). We then iterate over the
+     * in-memory SnapshotInfo instances first, without any further forking, and only once that's complete do we start to retrieve
+     * SnapshotInfo data from the repository which forks the per-blob tasks to SNAPSHOT_META. It's important that we don't do any more
+     * non-forking iteration once we've started this forking phase. Once all the per-repository iteration is complete, we fork back to
```

Comment:
> It's important that we don't do any more ...

somehow the why is missing, or maybe it is just the formulation.

Reply:
Does 87892e60cda help?